hls4ml/templates/vivado/nnet_utils/nnet_graph.h

```diff
@@ -0,0 +1,432 @@
+			typename CONFIG_T::dense_config1::weight_t  core_node_w0[CONFIG_T::dense_config1::n_in*CONFIG_T::dense_config1::n_out],
+			typename CONFIG_T::dense_config1::bias_t    core_node_b0[CONFIG_T::dense_config1::n_out],
+			typename CONFIG_T::dense_config2::weight_t  core_node_w1[CONFIG_T::dense_config2::n_in*CONFIG_T::dense_config2::n_out],
+			typename CONFIG_T::dense_config2::bias_t    core_node_b1[CONFIG_T::dense_config2::n_out],
+			typename CONFIG_T::dense_config3::weight_t  core_node_w2[CONFIG_T::dense_config3::n_in*CONFIG_T::dense_config3::n_out],
+			typename CONFIG_T::dense_config3::bias_t    core_node_b2[CONFIG_T::dense_config3::n_out],
+			typename CONFIG_T::dense_config4::weight_t  core_node_w3[CONFIG_T::dense_config4::n_in*CONFIG_T::dense_config4::n_out],
+			typename CONFIG_T::dense_config4::bias_t    core_node_b3[CONFIG_T::dense_config4::n_out])
+  {
+    //initialize arrays
+    //1. node_attr (input)
+    data_T node_attr[CONFIG_T::n_node][CONFIG_T::node_dim];
+    #pragma HLS ARRAY_PARTITION variable=node_attr complete dim=0
+    nnet::vec_to_mat<data_T, data_T, typename CONFIG_T::node_attr_config>(node_attr_1D, node_attr);
+
+    //2. edge_attr_aggr (input)
+    data_T edge_attr_aggr[CONFIG_T::n_node][CONFIG_T::edge_dim];
+    #pragma HLS ARRAY_PARTITION variable=edge_attr_aggr complete dim=0
+    nnet::vec_to_mat<data_T, data_T, typename CONFIG_T::edge_attr_aggr_config>(edge_attr_aggr_1D, edge_attr_aggr);
+
+    // 3. node_update (output)
+    res_T node_update[CONFIG_T::n_node][CONFIG_T::out_dim];
+    #pragma HLS ARRAY_PARTITION variable=node_update complete dim=0
+
+    #pragma HLS PIPELINE II=CONFIG_T::reuse_factor
+    node_loop: for(int i = 0; i < CONFIG_T::n_node; i++){ //for each node
+      #pragma HLS UNROLL
+
+      // construct NN input: <node, edge_attr_aggr>
+      data_T phi_input[CONFIG_T::edge_dim + CONFIG_T::node_dim];
+      #pragma HLS ARRAY_PARTITION variable=phi_input complete dim=0
+      nnet::concatenate1d<data_T, data_T, data_T, typename CONFIG_T::merge_config1>(node_attr[i], edge_attr_aggr[i], phi_input);
+
+      // send it through NN
+        if(CONFIG_T::n_layers == 1){
+	      nnet::dense_mult_1lyr<data_T, res_T, CONFIG_T>(phi_input, node_update[i], core_node_w0, core_node_b0);
+        }
+        else if(CONFIG_T::n_layers == 2){
+	      nnet::dense_mult_2lyr<data_T, res_T, CONFIG_T>(phi_input, node_update[i], core_node_w0, core_node_b0, core_node_w1, core_node_b1);
+        }
+        else if(CONFIG_T::n_layers == 3){
+	      nnet::dense_mult_3lyr<data_T, res_T, CONFIG_T>(phi_input, node_update[i], core_node_w0, core_node_b0, core_node_w1, core_node_b1, core_node_w2, core_node_b2);
+        }
+        else { // CONFIG_T::n_layers == 4
+	      nnet::dense_mult_4lyr<data_T, res_T, CONFIG_T>(phi_input, node_update[i], core_node_w0, core_node_b0, core_node_w1, core_node_b1, core_node_w2, core_node_b2, core_node_w3, core_node_b3);
+        }
+    }

```

Comment:
@thesps do you mean something similar to @vloncar's code generation for CNNs: https://github.com/fastmachinelearning/hls4ml/compare/master...vloncar:cnn_parallel

Reply:
I was thinking generally that these `EdgeBlock` and `NodeBlock` are NNs that can have any architecture. But this code hardcodes a few special cases. Those special cases might be the most common ones, but can't we generalise? e.g. is there any issue with using 5 layers, or an activation other than ReLU?

I'm not sure what's the best way forward, but it may make more sense if the `EdgeBlock` and `NodeBlock` are separate `HLSModel`s (now `Model`s), with the full flexibility over NN architecture and configuration that that implies. In this PR those blocks are '`Layer`s' but they are actually full NN models - of the types we already support.

Reply:
@vloncar I was also thinking this. The only thing is that the `core_node_w` and `core_node_b` inputs would have to be indexed in a complicated way, because each of the "layers" in the generalized MLP may not be of the same size. It should be possible to pass some parameter detailing each layer-size, and then cut up the overall input into the separate layers. 

@thesps The `NodeBlock` does indeed just pass the input through the relevant NN, but there is a bit more logic to the `EdgeBlock` than just the NN (i.e. constructing the NN-input). It may be possible to construct the NN for each "Block" as its own external function, and then call it inside of the block. The main challenge is to make each block calls its own, unique NN-function. I think we can pass each NN-function as a parameter to its respective block, but I'm not sure if Vivado would be able to synthesize this. 

One way around this is to basically split each Block into two functions - one for the Block-specific logic, and one for the NN logic - and then call those two functions in succession. Either way, I think this would require adding a new method: `VivadoWriter.write_NNs()`


I imagine that all of these options will affect the usage/latency numbers, so I'm currently trying to test them all out and see which is the fastest/cheapest.